File: src/libs/posix/src/unistd/syscall/sbrk.rs
// Copyright(c) The Maintainers of Nanvix.
// Licensed under the MIT License.

//==================================================================================================
// Imports
//==================================================================================================

use ::nvx::{
    mm,
    mm::{
        VirtualAddress,
        PAGE_ALIGNMENT,
    },
    pm,
    pm::ProcessIdentifier,
    sys::error::{
        Error,
        ErrorCode,
    },
};

//==================================================================================================
// Standalone Functions
//==================================================================================================///

///
/// # Description
///
/// The `sbrk()` increments the location of the program break by `size` bytes. The program break
/// defines the end of the process's data segment and is the address of the first location after the
/// end of the uninitialized data segment. Increasing the program break has the effect of allocating
/// memory to the process; decreasing the break deallocates memory. Calling `sbrk()` with a zero
/// size increment can be used to find the current location of the program break.
///
/// # Parameters
///
/// - `size`: Number of bytes to increment the program break.
///
/// # Returns
///
/// Upon successful completion, the `sbrk()` function returns the address of the start of the newly
/// allocated memory. Otherwise, it returns an error code.
///
pub fn sbrk(size: isize) -> Result<*mut u8, Error> {
    ::nvx::trace!("sbrk(): size = {}", size);
    static mut END: *mut u8 = mm::BREAK_BASE_RAW as *mut u8;

    // Check if querying the current program break.
    if size == 0 {
        return Ok(unsafe { END });
    }

    let old_end: *mut u8 = unsafe {
        let old_end: *mut u8 = END;
        let new_end: *mut u8 = END.offset(size);

        // Align the new end.
        let new_end: *mut u8 = mm::align_up(new_end as usize, PAGE_ALIGNMENT) as *mut u8;

        // Check wether we should allocate or free memory.
        if size > 0 {
            // Allocate memory.

            // Check if we would exceed the heap size.
            if new_end >= (mm::BREAK_BASE_RAW + mm::C_HEAP_SIZE) as *mut u8 {
                return Err(Error::new(ErrorCode::OutOfMemory, "out of memory"));
            }

            let pid: ProcessIdentifier = pm::getpid()?;

            // Allocate memory.
            nvx::mm::heap::map_range(
                pid,
                VirtualAddress::from_raw_value(old_end as usize),
                VirtualAddress::from_raw_value(new_end as usize),
            )?;
        } else {
            // Free memory.

            // Check if we would free memory below the heap base address.
            if new_end < mm::BREAK_BASE_RAW as *mut u8 {
                return Err(Error::new(ErrorCode::InvalidArgument, "invalid size"));
            }

            let pid: ProcessIdentifier = pm::getpid()?;

            // Free memory.
            nvx::mm::heap::unmap_range(
                pid,
                VirtualAddress::from_raw_value(new_end as usize),
                VirtualAddress::from_raw_value(old_end as usize),
            )?;
        }

        END = new_end;
        old_end
    };

    Ok(old_end)
}

File: src/tests/memory-c/sbrk.c
/*
 * Copyright(c) The Maintainers of Nanvix.
 * Licensed under the MIT License.
 */

//==================================================================================================
// Configuration
//==================================================================================================

// Enable sbrk().
#define _BSD_SOURCE 1

//==================================================================================================
// Imports
//==================================================================================================

#include <assert.h>
#include <unistd.h>

//==================================================================================================
// Standalone Functions
//==================================================================================================

// Tests whether we can allocate memory using `sbrk()`.
void test_sbrk(void)
{
    // Check that `sbrk()` returns a pointer to the end of the heap.
    void *ptr = sbrk(0);
    assert(ptr != (void *)-1);

    // Check that `sbrk()` can allocate memory.
    void *new_ptr = sbrk(4096);
    assert(new_ptr != (void *)-1);
    assert(ptr == new_ptr);

    // Check that program break is where we expect it to be.
    void *new_ptr2 = sbrk(0);
    assert(new_ptr2 != (void *)-1);
    assert(new_ptr2 == (void *)((char *)ptr + 4096));

    // Check that `sbrk()` can free memory.
    void *free_ptr = sbrk(-4096);
    assert(free_ptr != (void *)-1);
    assert(free_ptr == (void *)((char *)ptr + 4096));

    // Check that program break is where we expect it to be.
    void *new_ptr3 = sbrk(0);
    assert(new_ptr3 != (void *)-1);
    assert(new_ptr3 == ptr);
}
